dvc/render/data.py

```diff
@@ -0,0 +1,211 @@
+        return datapoints
+    assert isinstance(fields, set)
+
+    new_data = []
+    for data_point in datapoints:
+        keys = set(data_point.keys())
+        if keys & fields != fields:
+            raise FieldsNotFoundError(fields, keys)
+
+        new_data.append(
+            {key: value for key, value in data_point.items() if key in fields}
+        )
+
+    return new_data
+
+
+def _lists(dictionary: Dict):
+    for _, value in dictionary.items():
+        if isinstance(value, dict):
+            yield from _lists(value)
+        elif isinstance(value, list):
+            yield value
+
+
+def _find_first_list(data: Union[Dict, List], fields: Set) -> List[Dict]:
+    fields = fields or set()
+
+    if not isinstance(data, dict):
+        return data
+
+    for lst in _lists(data):
+        if (
+            all(isinstance(dp, dict) for dp in lst)
+            # if fields is empty, it will match any set
+            and set(first(lst).keys()) & fields == fields
+        ):
+            return lst
+
+    raise PlotDataStructureError()
+
+
+def _append_index(datapoints: List[Dict]) -> List[Dict]:
+    if INDEX_FIELD in first(datapoints).keys():
+        return datapoints
+
+    for index, data_point in enumerate(datapoints):
+        data_point[INDEX_FIELD] = index
+    return datapoints
+
+
+class Converter:
+    """
+    Class that takes care of converting unspecified data blob
+    (Dict or List[Dict]) into datapoints (List[Dict]).
+    If some properties that are required by Template class are missing
+    ('x', 'y') it will attempt to fill in the blanks.
+    """
+
+    @staticmethod
+    def update(datapoints: List[Dict], update_dict: Dict):
+        for data_point in datapoints:
+            data_point.update(update_dict)
+        return datapoints
+
+    def __init__(self, plot_properties: Optional[Dict] = None):
+        plot_properties = plot_properties or {}
+        self.props = deepcopy(plot_properties)
+        self.inferred_props: Dict = {}
+
+        self.skip_steps: Set[str] = set()
+
+        self._infer_x()
+        self._infer_fields()
+
+        self.steps = []
+        self.steps.append(
+            (
+                "find_data",
+                partial(
+                    _find_first_list,
+                    fields=self.inferred_props.get("fields", set())
+                    - {INDEX_FIELD},
+                ),
+            )
+        )
+        self.steps.append(("append_index", partial(_append_index)))
+        self.steps.append(
+            (
+                "filter_fields",
+                partial(
+                    _filter_fields,
+                    fields=self.inferred_props.get("fields", set()),
+                ),
+            )
+        )
+
+    def _infer_x(self):
+        if self.props.get("x", None):
+            self.skip_step("append_index")
+        else:
+            self.inferred_props["x"] = INDEX_FIELD
+
+    def skip_step(self, name: str):
+        self.skip_steps.add(name)
```

Comment:
Can simply drop from planned steps immediately
```suggestion
        self.steps = [(_name, fn) for _name, fn in self.steps if _name != name]
```